File: sentenceSimilarity734.cpp
/*
Given two sentences words1, words2 (each represented as an array of strings), and a list of similar word pairs pairs, determine if two sentences are similar.

For example, "great acting skills" and "fine drama talent" are similar, if the similar word pairs are pairs = [["great", "fine"], ["acting","drama"], ["skills","talent"]].

Note that the similarity relation is not transitive. For example, if "great" and "fine" are similar, and "fine" and "good" are similar, "great" and "good" are not necessarily similar.

However, similarity is symmetric. For example, "great" and "fine" being similar is the same as "fine" and "great" being similar.

Also, a word is always similar with itself. For example, the sentences words1 = ["great"], words2 = ["great"], pairs = [] are similar, even though there are no specified similar word pairs.

Finally, sentences can only be similar if they have the same number of words. So a sentence like words1 = ["great"] can never be similar to words2 = ["doubleplus","good"].

Note:

The length of words1 and words2 will not exceed 1000.
The length of pairs will not exceed 2000.
The length of each pairs[i] will be 2.
The length of each words[i] and pairs[i][j] will be in the range [1, 20].

wrong test case:
["great"]
["great"]
[]
*/
class Solution {
public:
    bool areSentencesSimilar(vector<string>& words1, vector<string>& words2, vector<vector<string>>& pairs) {
		if (words1.size() != words2.size()) return false;
		int count = 0;
 		for (int i = 0; i < words1.size(); i++) {
			if (words1[i] == words2[i]) {
				count++;
				continue;
			} else {
				for (int j = 0; j < pairs.size(); j++) {
					if (pairs[j][0] == words1[i]) {
						if (pairs[j][1] == words2[i]) {
							count++;
							break;
						}
					} else if (pairs[j][1] == words1[i]) {
						if (pairs[j][0] == words2[i]) {
							count++;
							break;
						}
					}
				}
			}
		}
		return count == words1.size();
    }
};
//the fatest method
class Solution {
public:
    bool areSentencesSimilar(vector<string>& words1, vector<string>& words2, vector<vector<string>>& pairs) {
        if(words1.size()!=words2.size()) return false;
        set <string> pairset;
        for(int i=0;i<pairs.size();i++)
        {
            string temp = pairs[i][0]+"+"+pairs[i][1];
            pairset.insert(temp);
        }
        for(int i=0;i<words1.size();i++)
        {
            if(words1[i]!=words2[i] && pairset.find(words1[i] + "+" + words2[i])==pairset.end() &&
                    pairset.find(words2[i] + "+" + words1[i])==pairset.end())
                return false;
        }
        return true;
    }
};

File: sentenceSimilarity734.py
"""
We can represent a sentence as an array of words, for example, the sentence "I am happy with leetcode" can be represented as arr = ["I","am",happy","with","leetcode"].

Given two sentences sentence1 and sentence2 each represented as a string array and given an array of string pairs similarPairs where similarPairs[i] = [xi, yi] indicates that the two words xi and yi are similar.

Return true if sentence1 and sentence2 are similar, or false if they are not similar.

Two sentences are similar if:

They have the same length (i.e., the same number of words)
sentence1[i] and sentence2[i] are similar.
Notice that a word is always similar to itself, also notice that the similarity relation is not transitive. For example, if the words a and b are similar, and the words b and c are similar, a and c are not necessarily similar.
"""
class Solution:
    def areSentencesSimilar(self, sentence1: List[str], sentence2: List[str], similarPairs: List[List[str]]) -> bool:
        if len(sentence1) != len(sentence2):
            return False
        sim = dict()
        for i in range(len(similarPairs)):
            if similarPairs[i][0] in sim.keys():
                sim[similarPairs[i][0]].add(similarPairs[i][1])
            else:
                sim[similarPairs[i][0]] = {similarPairs[i][1]}

        for i in range(len(sentence1)):
            if sentence1[i] != sentence2[i]:
                if sentence1[i] in sim.keys() and sentence2[i] in sim[sentence1[i]]:
                    continue
                elif sentence2[i] in sim.keys() and sentence1[i] in sim[sentence2[i]]:
                    continue
                else:
                    return False
        return True
        

